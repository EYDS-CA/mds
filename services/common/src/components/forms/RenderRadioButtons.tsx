import React, { FC, useContext } from "react";
import { Form, Radio } from "antd";
import { BaseInputProps, BaseViewInput, getFormItemLabel } from "@mds/common/components/forms/BaseInput";
import { IRadioOption } from "@mds/common/interfaces";
import { FormContext } from "./FormWrapper";

/**
 * @class RenderRadioButtons - Ant Design `Radio` component used for boolean values in redux-form.
 */

interface RenderRadioButtonsProps extends BaseInputProps {
  label: string;
  customOptions?: IRadioOption[];
  optionType?: "default" | "button";
  isVertical?: boolean;
}

const RenderRadioButtons: FC<RenderRadioButtonsProps> = ({
  meta,
  label,
  disabled = false,
  input,
  id,
  help,
  customOptions,
  labelSubtitle,
  required = false,
  optionType = "default",
  isVertical = false,
  showOptional = true,
}) => {
  const { isEditMode } = useContext(FormContext);

  const options = customOptions ?? [
    { label: "Yes", value: true },
    { label: "No", value: false },
  ];

  const handleRadioChange = (e) => {
    input.onChange(e.target.value);
  };

  if (!isEditMode) {
    if (optionType !== "default") {
      const matching = options.find((opt) => opt.value === input.value);
      return <BaseViewInput label={label} value={matching?.label} />
    }
    const radioGroupClass = isVertical ? "vertical-radio-group view-item" : "view-item"
    return (
      <Form.Item
        id={id}
        getValueProps={() => ({ value: input.value })}
        name={input.name}
        label={<div className="view-item-label">{getFormItemLabel(label, false, labelSubtitle, false)}</div>}
        className="view-item"
      >
        <>
          <Radio.Group
            disabled={true}
            name={input.name}
            value={input.value}
            options={options}
            optionType={optionType}
            className={radioGroupClass}
            buttonStyle="solid"
          />
          {help && <div className={`form-item-help ${input.name}-form-help`}>{help}</div>}
        </>
      </Form.Item>
    );
  }

  return (
    <Form.Item
      id={id}
      getValueProps={() => ({ value: input.value })}
      name={input.name}
      required={required}
      validateStatus={meta.touched ? (meta.error && "error") || (meta.warning && "warning") : ""}
      help={
        meta.touched &&
        ((meta.error && <span>{meta.error}</span>) || (meta.warning && <span>{meta.warning}</span>))
      }
      label={getFormItemLabel(label, required, labelSubtitle, showOptional)}
    >
      <>
        <Radio.Group
          disabled={disabled}
          name={input.name}
          value={input.value}
          onChange={handleRadioChange}
          options={options}
          optionType={optionType}
          buttonStyle="solid"
          {...(isVertical && { className: "vertical-radio-group" })}
        />
        {help && <div className={`form-item-help ${input.name}-form-help`}>{help}</div>}
      </>
    </Form.Item>
  );
};

export default RenderRadioButtons;
