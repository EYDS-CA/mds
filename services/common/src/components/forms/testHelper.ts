export const inputMeta = {
    touched: false,
    error: false,
    warning: false,
    autofilled: false,
    asyncValidating: false,
    dirty: false,
    dispatch: jest.fn(),
    form: "FORM",
    initial: undefined,
    invalid: false,
    pristine: true,
    submitFailed: false,
    submitting: false,
    valid: true,
    visited: false,
};

export const errorInputMeta = {
    touched: true,
    error: "Error message",
    warning: false,
    autofilled: false,
    asyncValidating: false,
    dirty: true,
    dispatch: jest.fn(),
    form: "FORM",
    initial: undefined,
    invalid: true,
    pristine: true,
    submitFailed: false,
    submitting: false,
    valid: true,
    visited: true,
}

// needs a value and a name
export const inputProps = {
    onBlur: jest.fn(),
    onChange: jest.fn(),
    onDragStart: jest.fn(),
    onDrop: jest.fn(),
    onFocus: jest.fn()
}