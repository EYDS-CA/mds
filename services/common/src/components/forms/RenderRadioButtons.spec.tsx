import React from "react";
import { render } from "@testing-library/react";
import RenderRadioButtons from "./RenderRadioButtons";
import { inputMeta, inputProps } from "./testHelper";

describe("RenderRadioButtons", () => {
    it("renders properly", () => {
        const { container } = render(<RenderRadioButtons
            id="radioInput"
            label="Mock Label"
            customOptions={[
                { label: "Opt 1", value: "opt1" },
                { label: "Opt 2", value: "opt2" }
            ]}
            input={{ value: "opt1", name: "radioInput", ...inputProps }}
            meta={inputMeta}
        />);
        expect(container).toMatchSnapshot();
    })
})