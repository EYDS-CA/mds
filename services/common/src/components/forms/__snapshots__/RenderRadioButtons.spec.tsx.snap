// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderRadioButtons renders properly 1`] = `
<div>
  <div
    class="ant-form-item ant-form-item-with-help"
  >
    <div
      class="ant-row ant-form-item-row"
    >
      <div
        class="ant-col ant-form-item-label"
      >
        <label
          class=""
          for="radioInput"
          title=""
        >
          <div>
            Mock Label
             
            <span
              class="form-item-optional"
            >
              Â (optional)
            </span>
          </div>
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-radio-group ant-radio-group-solid"
            >
              <label
                class="ant-radio-wrapper ant-radio-wrapper-checked ant-radio-wrapper-in-form-item"
              >
                <span
                  class="ant-radio ant-radio-checked"
                >
                  <input
                    checked=""
                    class="ant-radio-input"
                    name="radioInput"
                    type="radio"
                    value="opt1"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Opt 1
                </span>
              </label>
              <label
                class="ant-radio-wrapper ant-radio-wrapper-in-form-item"
              >
                <span
                  class="ant-radio"
                >
                  <input
                    class="ant-radio-input"
                    name="radioInput"
                    type="radio"
                    value="opt2"
                  />
                  <span
                    class="ant-radio-inner"
                  />
                </span>
                <span>
                  Opt 2
                </span>
              </label>
            </div>
          </div>
        </div>
        <div
          style="display: flex; flex-wrap: nowrap;"
        >
          <div
            class="ant-form-item-explain ant-form-item-explain-connected"
            id="radioInput_help"
            role="alert"
          >
            <div
              class=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
